File: kotlin/0118-e-pascals-triangle.kt
// https://leetcode.com/problems/pascals-triangle/
class Solution {
    fun generate(numRows: Int): List<List<Int>> {
        if (numRows == 1) return listOf(listOf(1))
        if (numRows == 2) return listOf(listOf(1),listOf(1,1))
        
        val pascal_tri = mutableListOf(listOf(1),listOf(1,1))

        for (i in 2 until numRows) {
            val inner = mutableListOf(1)
            val before = pascal_tri[i-1]

            for (j in 0 until before.size-1) {
                inner.add(before[j] + before[j+1])
            }
            inner.add(1)
            
            pascal_tri.add(inner)
        }

        return pascal_tri
    }
}

File: python/0118-e-pascals-triangle.py
# https://leetcode.com/problems/pascals-triangle/
from typing import List

class Solution:
    def generate(self, numRows: int) -> List[List[int]]:
        if numRows == 1:
            return [[1]]
        if numRows == 2:
            return [[1],[1,1]]
        
        pascal_tri = [[1],[1,1]]
        for i in range(2, numRows):
            inner = [1]
            before = pascal_tri[i-1]
            for j in range(0, len(before)-1):
                inner.append(before[j] + before[j+1])
            inner.append(1)
            pascal_tri.append(inner)
        
        return pascal_tri

# Runner code with sample input
if __name__ == "__main__":
    # Test cases: (numRows, expected_output)
    test_cases = [
        (1, [[1]]),
        (2, [[1], [1,1]]),
        (3, [[1], [1,1], [1,2,1]]),
        (4, [[1], [1,1], [1,2,1], [1,3,3,1]]),
        (5, [[1], [1,1], [1,2,1], [1,3,3,1], [1,4,6,4,1]]), # Edge case
    ]
    
    solution = Solution()
    
    # Test each case
    for i, (numRows, expected) in enumerate(test_cases, 1):
        result = solution.generate(numRows)
        print(f"Test Case {i}:")
        print(f"Number of rows: {numRows}")
        print(f"Expected triangle:")
        for row in expected:
            print(row)
        print(f"Got:")
        for row in result:
            print(row)
        print(f"{'✓ Passed' if result == expected else '✗ Failed'}")
        print("-" * 50)